sdk/eventgrid/azure-messaging-eventgrid/src/main/java/com/azure/messaging/eventgrid/EventGridPublisherAsyncClient.java

```diff
@@ -158,6 +171,21 @@ public EventGridServiceVersion getServiceVersion() {
     Mono<Response<Void>> sendCustomEventsWithResponse(Iterable<Object> events, Context context) {
         return Flux.fromIterable(events)
             .collectList()
-            .flatMap(list -> this.impl.publishCustomEventEventsWithResponseAsync(this.hostname, list, context));
+            .flatMap(list -> this.impl.publishCustomEventEventsWithResponseAsync(this.hostname, list,
+                context.addData(AZ_TRACING_NAMESPACE_KEY, Constants.EVENT_GRID_TRACING_NAMESPACE_VALUE)));
+    }
+
+    private void addCloudEventTracePlaceHolder(Iterable<CloudEvent> events) {
+        if (TracerProxy.isTracingEnabled()) {
+            for (CloudEvent event : events) {
+                if (event.getExtensionAttributes() == null ||
+                    (event.getExtensionAttributes().get(Constants.TRACE_PARENT) == null &&
+                    event.getExtensionAttributes().get(Constants.TRACE_STATE) == null)) {
+
+                    event.addExtensionAttribute(Constants.TRACE_PARENT, Constants.TRACE_PARENT_PLACEHOLDER);
+                    event.addExtensionAttribute(Constants.TRACE_STATE, Constants.TRACE_STATE_PLACEHOLDER);
```

Comment:
Do we need both?

Reply:
I put the two placeholder separately. After placeholder replacement, whatever in the request headers will be put into the body eventually. There won't be two in the header has just one.